### ISSUE TITLE
notnull constraint works in VS2022 but not in VS2026

### ISSUE THREAD

**Issue by hudecekd** [dotnet/roslyn#81654]
Hello,

I am using VS 2026 for development.
Currently we have found out that build is not working in Azure but works locally in VS 2026.
We have found out that when using dotnet cli build fails as expected but in VS 2026 it is successful. We have also found out that build in VS 2022 also fails which is expected.

**Steps to reproduce**
1. Open in VS 2022
2. Build app and it will fail as expected.
3. Open in VS 2026
4. Build app and it will work. - This is problem.

Here is minimal code to reproduce the problem.
[ConsoleApp147.zip](https://github.com/user-attachments/files/24103268/ConsoleApp147.zip)

**Expected Behavior**:
- CS8714 is reported when building in VS 2026.

**Actual Behavior**:
- CS8714 is not reported when building in VS 2026.
- CS8714 is reported when building in VS 2022.

**Comment by jjonescz** [comment]
Thanks for reporting this issue.

> dotnet cli build fails as expected

That seems to be because you have `global.json` pinning the SDK to .NET 8. If that is changed to .NET 10, the CLI doesn't fail either.

> ExecuteAsync is defined with "notnull" constraint.

This doesn't seem to be correct. There is no `notnull` constraint on the method: https://github.com/restsharp/RestSharp/blob/57284b9478a105ec07c9299cd9cf5a6c56350f07/src/RestSharp/RestClient.Extensions.cs#L41
So the behavior of .NET 10 compiler seems expected.

The error reported by the .NET 8 compiler seems to be due to a bug in how extension members are emitted by the .NET 10 compiler (the RestSharp library is compiled with the .NET 10 compiler). Specifically, consider the following code:

```cs
public static class E
{
    public static void X1(object o) { }
    public static void X2(object o) { }
    public static void M1<T>(this int i) { }
    extension(int i)
    {
        public void M2<T>() { }
    }
}
```

It gets emitted as:

```cs
[NullableContext(1)]
[Nullable(0)]
[Extension]
public static class E
{
	[SpecialName]
	[Extension]
	public sealed class <G>$BA41CFE2B5EDAEB8C1B9062F59ED4D69
	{
		[SpecialName]
		public static class <M>$F4B4FFE41AB49E80A4ECF390CF6EB372
		{
			[SpecialName]
			[CompilerGenerated]
			public static void <Extension>$(int i)
			{
			}
		}

		[NullableContext(2)]
		[ExtensionMarker("<M>$F4B4FFE41AB49E80A4ECF390CF6EB372")]
		public void M2<T>()
		{
			throw new NotSupportedException();
		}
	}

	public static void X1(object o)
	{
	}

	public static void X2(object o)
	{
	}

	[NullableContext(2)]
	[Extension]
	public static void M1<T>(int i)
	{
	}

	[Extension]
	public static void M2<T>(int i)
	{
	}
}
```

Note that there is `NullableContext(1)` on the extension container type (it gets lifted/shared there from `X1` and `X2` methods).
But there is no `NullableContext(2)` on the `M2` implementation method (this is the bug), so the nullable context from the type is "inherited" incorrectly, leading the old compiler thinking `M2` has the `notnull` constraint. (Whereas the new compiler probably uses the marker M2 method which has `NullableContext(2)` correctly.)

**Comment by hudecekd** [comment]
Hi, thanks for quick response @jjonescz ,
if I understand correctly it means that there is a but in .NET10 compiler which so when library compiled with .NET10 compiler is referenced in code which is built with .NET10 compile bug will not be presented.
But when it will be referenced in code which will be compiled with .NET8 compiler bug will present itsefl.
Right?

What should be correct approach to make it work.
Currently we have build in azure which fails so we fixed it by adding `notnull` constarint even if it should not be needed as discussed here. It is "hotfix". But I assume that similar bug can introduce  when other packages compiled in .NET10 compiled will be used.

So should we
- solve error by adding `notnull` as quit hotfix
- use .NET10 compiler to compile our .NET8 code which seems to ignore bug .NET10 introduces?

Thanks

**Comment by jjonescz** [comment]
> if I understand correctly it means that there is a but in .NET10 compiler which so when library compiled with .NET10 compiler is referenced in code which is built with .NET10 compile bug will not be presented.
> But when it will be referenced in code which will be compiled with .NET8 compiler bug will present itsefl.
> Right?

That sounds correct.

> So should we

You should use the .NET 10 compiler if possible, that seems like the best fix.

**Comment by hudecekd** [comment]
Hi, thanks for quick response @jjonescz ,
maybe i am missing something.

If i use global.json and fix build to .net8 sdk should not VS 2022 and vs 2026 both fail? Both should use same sdk for build right?
Am I missing something here?

And to clarify it. Bug is actually in .NET10 compiler right?

Thanks for reply.

**Comment by jjonescz** [comment]
> Both should use same sdk for build right?

It's more complicated. Before .NET 10, VS uses its own compiler version. Since .NET 10, the compiler that ships in the SDK should be always used instead (https://github.com/dotnet/sdk/blob/main/documentation/general/decouple-vs-and-net-sdk.md). Therefore, if you are on .NET 8, VS2022 uses its compiler version (an old one) and VS2026 uses its compiler version (a new one). (You can see which exact compiler is being used by placing `#error version` in some `.cs` file for example.)

> And to clarify it. Bug is actually in .NET10 compiler right?

Yes. But it only manifests if the library built with the buggy .NET10 compiler is consumed from .NET8 compiler. (In other words, if you only use .NET10, there should be no bug.)

**Comment by hudecekd** [comment]
Thnaks @jjonescz ,
I have wrongly assumed that VS uses global.json and then both VS should generate same results. Now it is clear.
We will update our cicd to use .net10 sdk for build to avoid future similar problems.

Thansk again.

**Comment by jjonescz** [comment]
> I have wrongly assumed that VS uses global.json and then both VS should generate same results

That will be a correct assumption since .NET 10 forward (it's a reasonable assumption, that's why we have changed the behavior to match).

VS actually always used global.json for some parts of the build (the MSBuild files) but not for C# compiler.
